daml-script/daml3/Daml/Script/Questions/Commands.daml

```diff
@@ -0,0 +1,102 @@
+-- Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE InstanceSigs #-}
+
+module Daml.Script.Questions.Commands where
+
+import Daml.Script.Internal
+
+data Command
+  = Create with
+      argC : AnyTemplate
+  | Exercise with
+      tplId : TemplateTypeRep
+      cId : ContractId ()
+      argE : AnyChoice
+  | ExerciseByKey with
+      tplId : TemplateTypeRep
+      keyE : AnyContractKey
+      argE : AnyChoice
+  | CreateAndExercise with
+      tplArgCE : AnyTemplate
+      choiceArgCE : AnyChoice
+
+-- The ledger api can only return create and exercise results, therefore `Exercise`, `ExerciseByKey` and `CreateByExercise` all return `ExerciseResult`s
+data CommandResult
+  = CreateResult (ContractId ())
+  | ExerciseResult LedgerValue
+
+-- | This is used to build up the commands send as part of `submit`.
+-- If you enable the `ApplicativeDo` extension by adding
+-- `{-# LANGUAGE ApplicativeDo #-}` at the top of your file, you can
+-- use `do`-notation but the individual commands must not depend
+-- on each other and the last statement in a `do` block
+-- must be of the form ``return expr`` or ``pure expr``.
+data Commands a = Commands with
+    commands : [Command]
+    continue : [CommandResult] -> a
+  deriving Functor
+
+-- Create and exercise gives 2 results, so we account for that here
+expectedCommandResults : Command -> Int
+expectedCommandResults CreateAndExercise {} = 2
+expectedCommandResults _ = 1
+
+-- Expected number of command results for a list of commands
+expectedCommandsResults : [Command] -> Int
+expectedCommandsResults = sum . fmap expectedCommandResults
```

Comment:
This has same complexity as length, but might iterate twice if `fmap` isn't lazy.
Consider a small refactor to merge sum and fmap.